packages/flutter/test/widgets/async_test.dart

```diff
@@ -0,0 +1,274 @@
+          findsOneWidget);
+    });
+  });
+  group('StreamFold', () {
+    testWidgets('gracefully handles transition from null stream',
+        (WidgetTester tester) async {
+      final GlobalKey key = new GlobalKey();
+      await tester.pumpWidget(new StringFold(key: key, stream: null));
+      expect(find.text(''), findsOneWidget);
+      final StreamController<String> controller =
+          new StreamController<String>();
+      await tester
+          .pumpWidget(new StringFold(key: key, stream: controller.stream));
+      expect(find.text('conn'), findsOneWidget);
+    });
+    testWidgets('gracefully handles transition to null stream',
+        (WidgetTester tester) async {
+      final GlobalKey key = new GlobalKey();
+      final StreamController<String> controller =
+          new StreamController<String>();
+      await tester
+          .pumpWidget(new StringFold(key: key, stream: controller.stream));
+      expect(find.text('conn'), findsOneWidget);
+      await tester.pumpWidget(new StringFold(key: key, stream: null));
+      expect(find.text('conn, disc'), findsOneWidget);
+    });
+    testWidgets('gracefully handles transition to other stream',
+        (WidgetTester tester) async {
+      final GlobalKey key = new GlobalKey();
+      final StreamController<String> controllerA =
+          new StreamController<String>();
+      final StreamController<String> controllerB =
+          new StreamController<String>();
+      await tester
+          .pumpWidget(new StringFold(key: key, stream: controllerA.stream));
+      await tester
+          .pumpWidget(new StringFold(key: key, stream: controllerB.stream));
+      controllerA.add('A');
+      controllerB.add('B');
+      await eventFiring(tester);
+      expect(find.text('conn, disc, conn, data:B'), findsOneWidget);
+    });
+    testWidgets('tracks events and errors until completion',
+        (WidgetTester tester) async {
+      final GlobalKey key = new GlobalKey();
+      final StreamController<String> controller =
+          new StreamController<String>();
+      await tester
+          .pumpWidget(new StringFold(key: key, stream: controller.stream));
+      controller.add('1');
+      controller.addError('bad');
+      controller.add('2');
+      controller.close();
+      await eventFiring(tester);
+      expect(
+          find.text('conn, data:1, error:bad, data:2, done'), findsOneWidget);
+    });
+  });
+}
+
+Future<Null> eventFiring(WidgetTester tester) async {
+  await tester.pump(const Duration(milliseconds: 0));
+}
+
+class StringFold extends StreamBuilderBase<String, List<String>> {
+  StringFold({Key key, Stream<String> stream})
+      : super(key: key, stream: stream);
+
+  @override
+  List<String> initial() => <String>[];
+
+  @override
+  List<String> afterConnected(List<String> current) => current..add('conn');
+
+  @override
+  List<String> afterData(List<String> current, String data) =>
+      current..add('data:$data');
```

Comment:
We'd probably put this all on one line.  We don't worry too much about hitting 80 cols